Update README.md file (#31)
v3io/scaler: README.md

# Scale to Zero

Infrastructure to scale any resource to and from zero

## Prerequisites

**Custom metrics API implementation:**

> The Autoscaler takes decisions based on data queried from Kubernetes 

Reviewer: ```suggestion
The Autoscaler makes decisions based on data queried from Kubernetes 
```

Reviewer: Move this below the paragraph explaining `**The autoscaler**` because you're "using" it here without defining it first, so the reader won't know what you're talking about yet

[custom metrics API.](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md)
There are several possible tools that implement it, we internally use
[Prometheus](https://prometheus.io/) with the [Prometheus-Adapter](https://github.com/DirectXMan12/k8s-prometheus-adapter)

> but you can use which ever you want... you can find [here](https://github.com/kubernetes/metrics/blob/master/IMPLEMENTATIONS.md#custom-metrics-api) 

Reviewer: ```suggestion
but you can use which ever you want! You can find some recommended implementations [here](https://github.com/kubernetes/metrics/blob/master/IMPLEMENTATIONS.md#custom-metrics-api) 
```

recommended implementations

## Design:
![Design scheme](https://www.lucidchart.com/publicSegments/view/cc8927a6-537f-4fe6-95e1-731503bc7996/image.jpeg)

> **The autoscaler** - responsible to periodically check whether resources should be scale to zero, it does that by 

Reviewer: ```suggestion
**The autoscaler** - Responsible to periodically check whether some resources should be scaled to zero, it does so by 
```

Suggest to add first a very generic definition of what is a "resource" like you suggested verbally. Should be first

> querying the custom metrics API. When it decide a resource should be scaled to zero it tells the resource scaler to 

Reviewer: ```suggestion
querying the custom metrics API. Upon deciding a resource should be scaled to zero, it uses the internal resource scaler to 
```

scale the resource to zero.

> The resource scaler will route the traffic to the DLX, which in terms of K8s is done by changing the svc selector, 

Reviewer: ```suggestion
The resource scaler will first route all incoming traffic to the DLX, which in terms of K8s is done by changing a service selector, 
```

> after that, it will scale the resource to zero.

Reviewer: ```suggestion
after doing so, it will perform the actual scaling to zero of the resource (by modifying a deployment / replicaset / daemon set / pods / etc).
```

> **The DLX** - responsible for receiving and buffering requests of scaled to zero resources, when it gets a message it 

Reviewer: ```suggestion
**The DLX** - Responsible for receiving and buffering requests on behalf of scaled to zero resources. When receiving a message addressed to the scaled to zero resource, it
```

> creates a buffer for the messages, and tells the resource scaler to scale the service back from zero.

Reviewer: ```suggestion
creates a buffer for the messages, and asks the resource scaler to scale the service back up from zero.
```

> The resource scaler will scale the resource back up and then route the traffic back to the service (by changing the svc 

Reviewer: ```suggestion
The resource scaler will scale the resource back up and then route the traffic back to the service (by changing the k8s service 
```

selector).

## Getting Started

> This infrastructure designed to be generic and extendable, it can scale any resource.

Reviewer: ```suggestion
The infrastructure is designed to be generic, flexible and extendable, so as to serve any resource we'd wish to scale to/from zero.
```

Reviewer: ☝️

> All you have to do is implement the resource scaler, the interface between it and this infrastructure's components is 

Reviewer: ```suggestion
All you have to do is implement the specific resource-scaler for your resource. The interface between your resource-scaler and the scale-to-zero infrastructure's components is 
```

Reviewer: If you're asking the "user" to implement component X, then introduce component X like you introduced the scaler explaining what is it's function in the flow, also to avoid ambiguities maybe it's best to not refer to it as `resource scaler` but `resource-scaler` or `resourcescaler` everywhere so it will be read as a single term (the word "scale" appear a lot here)

defined in [scaler-types](https://github.com/v3io/scaler-types)

> **Note:** Incompatibility between the scaler repo vendor dir and the resource scale vendor dir may break things, 

Reviewer: ```suggestion
**Note:** Incompatibility between this scaler vendor dir and your resource-scaler vendor dir may break things, 
```

> therefore it's suggested to put the resource scaler in its own repo

Reviewer: ```suggestion
therefore it's suggested to put your resource-scaler in its own repo
```

Examples:
* [Nuclio functions resource scaler](https://github.com/nuclio/nuclio/blob/master/pkg/platform/kube/resourcescaler/resourcescaler.go)
* [Iguazio's app service resource scaler](https://github.com/v3io/app-resource-scaler/blob/development/resourcescaler.go)  

## Installing

> Go plugin is the magic that glues the resource scaler and this infrastructure components together

Reviewer: ```suggestion
[Go plugins](https://appliedgo.net/plugins/) is the magic that glues the resource scaler and this infrastructure components together
```

> First you'll need to build a Dockerfile that builds your resource scaler as a Go plugin, and transplant it in this 

Reviewer: - Does it have to be a dockerfile? 🤷‍♂ if the user just builds a go binary why is that not enough?
- Transplant is not clear in this context imo. If I read this I wouldn't know what to do without going to the examples. Explain in a few sentences

Author: The whole context here is Kubernetes, in the design chart It's written Autoscaler **pod** etc... If it's a pod it's an image, and if it's an image there's a dockerfile...
It is possible to run it as a binary, but that doesn't make much sense.

Reviewer: it's all true but a stretch to the reader to understand all that from the word "transplant" :)

Infrastructure to scale any resource to and from zero

## Design:
![Design scheme](https://www.lucidchart.com/publicSegments/view/b01741e1-cb15-438e-8bd7-e5865dcc2043/image.jpeg)

**Resource** - A generic name for the component which is being scaled by this mechanism.
Will mostly include pod/s (that will be scaled when needed) wrapped by one or many deployment/daemon set/replica set, 
and k8s service/s that can be used to route incoming requests.

**Resource-scaler** - An implementation of the `ResourceScaler` interface defined in 
[scaler-types](https://github.com/v3io/scaler-types). It is transplanted inside the autoscaler and dlx using
[Go plugins](https://appliedgo.net/plugins/). They use them to perform actions on the specific resource.<br>
For example, when the autoscaler decides it needs to scale some resource to zero, it executes the resource-scaler's
`SetScale` function which has the knowledge how to scale to zero its specific resource.

> **The autoscaler** - Responsible to periodically check whether some resources should be scaled to zero, it does so by 

Reviewer: `Responsible for periodically checking whether some resources should be scaled to zero. This is performed by `

> querying the custom metrics API. Upon deciding a resource should be scaled to zero, it uses the internal resource-scaler 

Reviewer: `querying the custom metrics API. Upon deciding a resource should be scaled to zero, it uses the internal resource-scaler module ...`

to scale the resource to zero.
The resource-scaler will first route all incoming traffic to the DLX, which in terms of K8s is done by changing a 
service selector, after that, it will scale the resource to zero.

> **The DLX** - responsible for receiving and buffering requests of scaled to zero resources, when it gets a message it 

Reviewer: `responsible` -> `Responsible
`, , when it gets a message it ...` -> `. Upon receiving an incoming request it ...`

creates a buffer for the messages, and tells the resource-scaler to scale the service back from zero.

> The resource-scaler will scale the resource back up and then route the traffic back to the service (by changing the svc 

Reviewer: by modifying the k8s service selector

but you can use which ever you want! You can find some recommended implementations 
[here](https://github.com/kubernetes/metrics/blob/master/IMPLEMENTATIONS.md#custom-metrics-api)

## Getting Started
This infrastructure designed to be generic and extendable, it can scale any resource.
All you have to do is implement the specific resource-scaler for your resource. The interface between your 
resource-scaler and the scale-to-zero infrastructure's components is defined in 
[scaler-types](https://github.com/v3io/scaler-types)

**Note:** Incompatibility between this scaler vendor dir and your resource-scale vendor dir may break things, 
therefore it's suggested to put your resource-scaler in its own repo

Examples:
* [Nuclio functions resource-scaler](https://github.com/nuclio/nuclio/blob/master/pkg/platform/kube/resourcescaler/resourcescaler.go)
* [Iguazio's app service resource-scaler](https://github.com/v3io/app-resource-scaler/blob/development/resourcescaler.go)  

## Installing
[Go plugins](https://appliedgo.net/plugins/) is the magic that glues the resource-scaler and this infrastructure 
components together.<br>

> First you'll need to build the resource scaler as a Go plugin, for example: <br>

Reviewer: `resource scaler` -> `resource-scaler` everywhere then :) (since you went with this suggestion in some places make it consistent)

`GOOS=linux GOARCH=amd64 go build -buildmode=plugin -a -installsuffix cgo -ldflags="-s -w" -o ./plugin.so ./resourcescaler.go`<br>
The autoscaler/dlx looks for the plugin using this path (from the execution directory) `./plugins/*.so` so you should 

> move the outcome of the build command (the `plugin.so` file) to the `plugins` directory

Reviewer: `outcome` -> `binary artifact`

> It is much easier to do everything using a Dockerfile, here are some great examples: 

Reviewer: `a Dockerfile` -> `Dockerfiles`
